llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp

```diff
@@ -3287,6 +3287,76 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
     setOriginForNaryOp(I);
   }
 
+  // Convert `Mask` into `<n x i1>`.
+  Constant *createDppMask(unsigned Width, unsigned Mask) {
+    SmallVector<Constant *, 4> R;
+    R.assign(Width, ConstantInt::getFalse(F.getContext()));
+    for (auto &M : R) {
+      if (Mask & 1)
+        M = ConstantInt::getTrue(F.getContext());
+      Mask >>= 1;
+    }
+    return ConstantVector::get(R);
+  }
+
+  // Calculate output shadow as array of booleans `<n x i1>`, assuming if any
```

Comment:
"assuming" is a bit weird. Isn't that how the poison should propagate, not an assumption?

Reply:
Not exactly. Input bits can affect particular output bits.
E.g. argument with single uninitialized bit can be multiplied by initialized zero.
So it's approximation, but better checking will be expensive.

Reply:
Ah, ok. Please add a comment. I guess these instructions are beyond the reach of the C++ standard, so there is no proper definition for "undefined bits" that would say whether multiplying one by zero is UB or not?

Reply:
unfortunately have have to do such shortcuts even for C++ generated code

Reply:
OK